### ISSUE TITLE
Fl_Scroll with real *_BOX draws over scrollbars at non-default scales

### ISSUE THREAD

**Issue by dannye** [fltk/fltk#1149]
**Describe the bug**

When a Fl_Scroll has a "real" boxtype (instead of no box or a frame), then the scrollbars are not drawn properly at non-default scaling factors. The bug happens at all scales besides 100%, 200%, and 300%.

**To Reproduce**

```cpp
#include <FL/Fl.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Scroll.H>
#include <FL/Fl_Box.H>

int main(int argc, char **argv) {
	Fl_Double_Window *window = new Fl_Double_Window(100, 100, 400, 400);
	Fl_Scroll *scroll = new Fl_Scroll(50, 50, 300, 300);
	Fl_Box *box1 = new Fl_Box(-150, -150, 0, 0);
	Fl_Box *box2 = new Fl_Box(500, 500, 0, 0);
	window->end();

	scroll->color(fl_gray_ramp(16));
	scroll->box(FL_DOWN_BOX);

	window->show();
	return Fl::run();
}
```

1. Build the above demo.cxx and run.
2. Scale the window up or down.
3. Scroll left/right or up/down.
4. See that different parts of the scrollbars are drawn over by the scroll's box_.
5. Change `FL_DOWN_BOX` to `FL_DOWN_FRAME`.
6. See that the bug goes away.

**Expected behavior**

The scrollbars should always be completely drawn on top of the real box.

**Video**

https://github.com/user-attachments/assets/77cb76b7-bbed-4ba1-af31-5dd61a2f1d2f

**FLTK Version**
 - Version: 1.4.0
 - Downloaded and self-built from: Git
 - If from Git, branch: master
 - If from Git, commit: 12ea6a0

**FLTK Configure / Build Options**
 - Default VS CMake options

**Operating System / Platform:**
 - OS: Windows
 - OS Version: Windows 11
 - Processor if relevant: Intel

**Additional context**

The docs for Fl_Scroll include:

>   If all of the child widgets are packed together into a solid
  rectangle then you want to set box() to FL_NO_BOX or
  one of the _FRAME types. This will result in the best output.
  However, if the child widgets are a sparse arrangement you must
  set box() to a real _BOX type.

So a real boxtype is definitely meant to be valid.

**Comment by erco77** [comment]
Hmm, the redraw problem seems to be related to Fl_Scroll::draw()'s damage() handling.

There's some new code introduced by Manolo in a 2017 commit that looks odd to me.

I made this change to remove it which _seems_ to solve the redraw problem, but I'm not sure if this breaks something else, or perhaps is just removing some code that was used for testing. @ManoloFLTK might want to look at this one.

```diff
diff --git a/src/Fl_Scroll.cxx b/src/Fl_Scroll.cxx
index e32c5bda5..5c1dcafff 100644
--- a/src/Fl_Scroll.cxx
+++ b/src/Fl_Scroll.cxx
@@ -351,7 +351,7 @@ void Fl_Scroll::draw() {
   uchar d = damage();
 
   float scale = Fl_Surface_Device::surface()->driver()->scale();
-  if ((d & FL_DAMAGE_ALL) || scale != int(scale)) { // full redraw
+  if ((d & FL_DAMAGE_ALL)) { // full redraw
     draw_box(box(),x(),y(),w(),h(),color());
     draw_clip(this, X, Y, W, H);
   } else {
```
If that should be removed, then I think "float scale = .." can be removed too.

**Comment by erco77** [comment]
BTW, just to rule out the zero sized width of the two boxes inside the scroll possibly being a contributing factor, I made this modification to the OP's demo program to give the boxes a 50 x 50 size and make them visible (red + green). It doesn't affect the bug which still presents itself, so that shouldn't be the issue:

```
#include <FL/Fl.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Scroll.H>
#include <FL/Fl_Box.H>

int main(int argc, char **argv) {
        Fl_Double_Window *window = new Fl_Double_Window(100, 100, 400, 400);
        Fl_Scroll *scroll = new Fl_Scroll(50, 50, 300, 300);
        Fl_Box *box1 = new Fl_Box(-150, -150, 50, 50); box1->box(FL_FLAT_BOX); box1->color(FL_RED);
        Fl_Box *box2 = new Fl_Box(500, 500, 50, 50);   box2->box(FL_FLAT_BOX); box2->color(FL_GREEN);
        window->end();

        scroll->color(fl_gray_ramp(16));
        scroll->box(FL_DOWN_BOX);

        window->show();
        return Fl::run();
}
```

**Comment by erco77** [comment]
I'm going to leave this assigned to myself and Manolo, but await Manolo's input, as I'm not sure what the above mentioned 'scale' code does here, as this issue seems related to the scaling retrofit.

Adding myself to this issue because I did a bit of work on the handling of scrollbar calculations back in 2014 which seemed possibly related, in case that was a contributing factor. _(Not sure it is at this point, just following through)_

**Comment by dannye** [comment]
If the `scale != int(scale)` is important, then this would fix it:

```diff
diff --git a/src/Fl_Scroll.cxx b/src/Fl_Scroll.cxx
index e32c5bda5..769f88728 100644
--- a/src/Fl_Scroll.cxx
+++ b/src/Fl_Scroll.cxx
@@ -424,7 +424,7 @@ void Fl_Scroll::draw() {
   }
 
   // draw the scrollbars:
-  if (d & FL_DAMAGE_ALL) {
+  if ((d & FL_DAMAGE_ALL) || scale != int(scale)) {
     draw_child(scrollbar);
     draw_child(hscrollbar);
     if (scrollbar.visible() && hscrollbar.visible()) {
```

**Comment by erco77** [comment]
Dannye, regarding your recent comment:

Unless I'm missing something, your suggestion is just putting the current 1.4.0 code back into place, which seems to be causing the bug?

I was suggesting _removing_ the `|| scale != int(scale)` bit is what solves the bug.. or at least seems to in my quick tests.

**Comment by dannye** [comment]
Sorry, I mean; Repeat the `scale != int(scale)` condition on line 427 so that the condition is identical to the one on line 354.

**Comment by Albrecht-S** [comment]
@erco77 wrote:
> my understanding of the scaling stuff is it should all be hidden in the internals ...

I fully agree and I looked at the commit that introduced this code. It was a very, very big commit and I can imagine that this particular scaling code was an attempt to fix something **at that time** or it was a left-over from before the code had stabilized.

I'm really interested to know what @ManoloFLTK can tell us about this.

If we can't find a **real** reason why it must be there, then I propose to remove it and wait for other bug reports (if any). IMHO it's better to remove questionable code than to leave it.

**Comment by ManoloFLTK** [comment]
TLDR: don't remove `|| scale != int(scale)` ! It's here for a reason.

The effect of this debated statement
`  if ((d & FL_DAMAGE_ALL) || scale != int(scale)) { // full redraw`
is to perform a full widget redraw more often than only when `d & FL_DAMAGE_ALL` is true, rather than call `fl_scroll()` which scrolls up or down part of the window content by a few pixels. Such a change cannot produce inaccurate graphics because a full redraw is always correct. At worst, it slows down the program. In this sense, I believe it's not fair to call it _questionable code_.

Another issue is that the part of function `Fl_Scroll::draw()` which draws the scroll panel and that which draws the scrollbars must either both undo or both use the debated statement, otherwise one part may cancel the scrollbars while the other doesn't redraw them. This is what causes the present issue, and the fix is exactly what @dannye proposed.

Ultimately, the reason of this statement is that, at least with legacy, integer coordinates-only graphics systems (here GDI and X11-noCairo) and when the scaling factor (`s`) doesn't have an integral value, an element of the scroll area of an Fl_Scroll widget with ordinate `y` in FLTK units is drawn at a pixel ordinate that approximates the floating value `y * s`. Then function `fl_scroll()` scrolls down, say, the scroll area by `n` pixels. The same element of the scroll area may not find itself exactly at the same ordinate it would have been given if the widget had been fully redrawn.
The same effect occurs with the abcissa of elements of the scroll area. Overall, when the scaling factor is not 100% or 200%, the graphics output after function `fl_scroll()` is not pixel-accurate as it must be.

To see this, comment out the 2 instances of `|| scale != int(scale)` at lines 354 and 427 of function `Fl_Scroll::draw()` in file `src/Fl_Scroll.cxx`. Then run FLTK test program `pack` under X11, scale it up to 120%. Click once the down arrowhead of the vertical scroller and once the right arrowhead of the horizontal scroller: the right part of the scroll area is not pixel accurate.

This problem when `fl_scroll()` is applied and the GUI is scaled to a non integral value seems to occur at least partially also with modern, floating coordinates, graphics systems (Quartz and Cairo with X11 or with Wayland). Under macOS, for example, run test/pack with the `Fl_Scroll:draw()` function modified as written above, scale to 120%, scroll the panel up and down by sliding the vertical scrollbar repeatedly: the scroll area progressively becomes seriously broken.

I believe FLTK should fix the 2nd part of `Fl_Scroll::draw()` to allow version 1.4.1 to draw correctly Fl_Scroll widgets with any type of box (committed at 557a5e4). Later, a detailed study of function `fl_scroll()` under the 2 types of graphics systems (modern or legacy) can be performed by a volunteer and may be some improvements can be identified and implemented.

I will also add a warning notice to the documentation of function `fl_scroll()` that it may not be always exact when the scaling factor value is not a multiple of 100% and that using the Fl_Scroll widget instead is generally the good thing.

To be complete, there may be yet another issue that Albrecht reported in #1138:
>  It looks as if the Fl_Scroll widget doesn't have a border and the children (boxes inside the Fl_Pack) draw over the border and thus leave such anti-aliasing artifacts at or outside the border.

This issue is not completely identified, at least in my mind, but seems unrelated to the use of function `fl_scroll()`. @Albrecht-S: You wrote _I'll investigate further_. Any new finding?

**Comment by Albrecht-S** [comment]
@ManoloFLTK You wrote:
> I believe it's not fair to call it questionable code.

I apologize if you understood this in a negative sense, there was no offense intended. Please see it in a neutral sense that this code needed to be checked.

Thank you for the detailed description why this code is required. After reading this I understand that we need it.

My reasoning is now: `fl_scroll()` is problematic as you explained because it can't always be pixel accurate when used with non-integral scaling. I imagine images being displayed with scaling and partially moved around. This is new in FLTK 1.4 and was obviously not a problem when `fl_scroll()` was created. OK; that's how it is. I doubt that we can "fix" `fl_scroll()` but I would be positively suprised if we could but that can be investigated later.

Therefore I agree that drawing the whole widget for non-integral scaling factors is the right compromise, at least for now. Thanks for your commit.

PS: Regarding issue 1138: no news yet, I've been busy with other things.

**Comment by Albrecht-S** [comment]
@dannye Please close this issue if commit 557a5e4fd0e379817a54ea0c16b62d7bd4aa086a fixes the issue for you. TIA.

**Comment by dannye** [comment]
Thanks @ManoloFLTK, I appreciate your expertise.

**Comment by erco77** [comment]
Thanks Manolo.

My only suggestion is to include a one line code comment with a url linking to manolo's description for future reference. e.g.

```diff
+   // See: https://github.com/fltk/fltk/issues/1149#issuecomment-2511135611
    float scale = Fl_Surface_Device::surface()->driver()->scale();
```

**Comment by ManoloFLTK** [comment]
> include a one line code comment with a url …

done at 8f8533c.
